sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala

```diff
@@ -180,10 +180,20 @@ abstract class OffsetWindowFunctionFrameBase(
     }
   }
 
+  /** Indicates whether the default values are Literal values. */
+  protected lazy val onlyLiteralNulls = expressions.forall { e =>
+    e.default == null || (e.default.foldable && e.default.eval() == null)
+  }
   override def prepare(rows: ExternalAppendOnlyUnsafeRowArray): Unit = {
     resetStates(rows)
     if (absOffset > rows.length) {
-      fillDefaultValue(EmptyRow)
+      // Avoid evaluating non-literal defaults with EmptyRow,
+      // which causes NullPointerException.
+      // Check whether defaults are Literal or null.
+      if (onlyLiteralNulls) {
```

Comment:
I think we should apply `fillDefaultValue` for any foldable expression here, not only the null literal.